core/src/main/java/org/apache/iceberg/io/RollingDataWriter.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.io;
+
+import java.util.List;
+import org.apache.iceberg.DataFile;
+import org.apache.iceberg.FileFormat;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.encryption.EncryptedOutputFile;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+
+/**
+ * A rolling data writer that splits incoming data into multiple files within one spec/partition.
+ */
+public class RollingDataWriter<T> extends RollingFileWriter<T, DataWriter<T>, DataWriteResult> {
+
+  private final WriterFactory<T> writerFactory;
+  private final List<DataFile> dataFiles;
+
+  public RollingDataWriter(WriterFactory<T> writerFactory, OutputFileFactory fileFactory,
+                           FileIO io, FileFormat fileFormat, long targetFileSizeInBytes,
+                           PartitionSpec spec, StructLike partition) {
+    super(fileFactory, io, fileFormat, targetFileSizeInBytes, spec, partition);
+    this.writerFactory = writerFactory;
+    this.dataFiles = Lists.newArrayList();
+    openCurrent();
+  }
+
+  @Override
+  protected DataWriter<T> newWriter(EncryptedOutputFile file) {
+    return writerFactory.newDataWriter(file, spec(), partition());
```

Comment:
The current `BaseRollingWriter` implementation passes the partition key to this method each time. It looks like the new classes expose `spec()` and `partition()` to avoid passing them here, since these 3 method implementations are the only place where those getters are called. I think I'd prefer dropping the 2 getter methods and just passing the spec and partition in here. What do you think?

Reply:
Alternatively, these implementations could keep their own copies of `spec` and `partition` in the constructor to avoid it. But I'd rather not do that.

Reply:
I did not pass the spec and partition on purpose as our rolling writers, as opposed to other more sophisticated writers, can write only to a single partition. If I see the spec and partition being passed, I won't know it is always the same value.

Reply:
I don't have a strong opinion here. If you feel it is not worth it, I can surely adapt.